### ISSUE TITLE
Flickering when adding a new polygon entity

### ISSUE THREAD

**Issue by hpinkos** [AnalyticalGraphicsInc/cesium#3945]
Reported on the forum: https://groups.google.com/forum/?hl=en#!topic/cesium-dev/xKVI49Ux430

In this example, the polygons flicker each time a new polygon is added to the collection.  It looks like the polygons are drawn solid instead of with an alpha for a split second.

``` javascript
var viewer = new Cesium.Viewer('cesiumContainer');

var dataSource = new Cesium.CustomDataSource('Polygons');
viewer.dataSources.add(dataSource);

var x = 20;
var y = -120;

setInterval(function(){
   x -= 5;
   y += 2; 

    var area = [0 + y, 51 + x, 5 + y, 51 + x, 2.50 + y, 48 + x];
    dataSource.entities.add({
        name: 'test',
        show : true,
        polygon : {
            hierarchy :{
                positions: Cesium.Cartesian3.fromDegreesArray(area)
            },
            material : Cesium.Color.RED.withAlpha(0.5)
        }
    });
}, 1000);
```

**Comment by denverpierce** [comment]
I can confirm this on time-dynamic entities as well.  If an entity goes from transparent to solid, such as in a time interval property color, there's about one frame where the transparent color is rendered as solid.

**Comment by dfil** [comment]
One interesting thing to note: using a CallbackProperty to update the position instead does stop the flickering (but is extremely slow due to recalculating all the geometry every frame).

**Comment by mertkurkcuoglu** [comment]
Hi cesium team; 
I want to inform you that flickering does not seen in primitive rendering. I tested it for adding several polylines as an entity at first, i encountered flickering problem while moving on any polyline. Then when i switch entity to primitive, flickering problem has been lost.

**Comment by hpinkos** [comment]
Thanks for the extra information @mertkurkcuoglu!

**Comment by jeremywilliams** [comment]
Hi cesium team, 

I am still experiencing this issue. Whenever I click on a polygon shape the screen flickers. I was wondering if anyone has a workaround or knows of a fix in the works?

**Comment by hpinkos** [comment]
Hello @jeremywilliams.  Can you please post a short [Sandcastle](http://cesiumjs.org/Cesium/Apps/Sandcastle/index.html) example that reproduces the issue you're seeing?  Thanks!

**Comment by roderickgreen** [comment]
We are also experiencing this. Any help would be appreciated. [Here](https://cesiumjs.org/Cesium/Build/Apps/Sandcastle/#c=rVbbbts4EP0Vwi+RUYWW4wYFXDeo43jRoE4b1NktFutiQUtji7sUKZCUE7fwv5ek7r5kg0X9ZJ2ZM+Sc4aG04Bsi0YbCI0j0DnF4RBNQNEvwHw7zzkL3OBFcE8pBnvnox4IjpGkCzABDtCJMgW8xwmlCNBW8CSpgocVueURDooVsBilfiWvx1IRWGWMqlAD8OtO6XWsNIhQRtEosiYIZ2YK8p+G/7ZAKgcOdIQzLpuYlgueT6afpxU2+i4SsQW7vpdjQuvqC77pvFzzXBrtaeM3EErBdciKYsIoVhd0j/vrh9mFqOAunqoRoyhhNFZi8ogxwTTUFhUkUeU7IVCiaa1aWIlKbf4QP8EqK5AbWRgzlnfeDAQ589DrAQddtm5PEdHb2BSIExTqCI5XJFQnhzKUU+DCfGUIx0HWshyjw82cFCb2jXMjxE1VDdHEZmB9uRck/ZfR10IqaUYOkhA3bGnyZ3uBHquMxS2PiBXjQtem7Qs1cmSXL4F6w7drs9xlpigavTbaRKU/f77DAT3YYU5BEhvH2P/QdS0m23l85CaHzfv/Sqj14U/XbAC9aYPDGgYM2eOHA/jHQFMmxb93ntbye/T59VkxvlXHnLuMkpQljHwiPmLFtN9fDih3nUNvdc+eweWp0nG6M8iWvddpDYqal8pVQWQcr0Lc8zfTYLVzvwION7pZjsAuHpdD1jEMzUUlwaqyaW0PQyPJw6QIfHfEblKl2K7Y6XSGvql6taS6gKCqO1Vg3EgrWzmnnn5LgYZsCnk1/e/h7MrudfCzbrvpzxSnXrdJVv6fOsP3VFq+Ifh0zJYd1CwiFdvR7B+HP6Wz2+eveUfBrTkqfgM3pd2OXflDCuyJhlwuwO9bPoVj7ExRrSdKYho27roE6C1WOOtTcVmFE1+Q7omOsRXmrNVfAJpHqLIIW2V0RLyELvj5gQ6oMu1+M8tkx7V8kh1fJ/75MkO3i3OzFt1q4P8dDrw5Cr14QsvQy8q0+FSfulKNHqWGQ4rycODETKkMGv9YC+y+p49off1nhg4T6fXWY8HJJLruViw5NtOt6uUTV18F3IZIH4e1JYJM6fmek9JbBVV7nPU1SITXKJPMw7mlIUjNCUL1lZj5fNA6Vo416JWkU0Q2i0btFZ+9LbNFBISNKmYj9ZrLuX3SuRj2T36IxQSLK1583IBnZ2pS4fzXLQYzxqGceD1laCLYkslHxJw) is a sandcastle demo. When you click in the scene it adds a new polygon entity and the others flicker.

**Comment by hpinkos** [comment]
@roderickgreen unfortunately we don't have a workaround for this, and I'm not sure how soon we'll be able to look into fixing it.  If you have some extra time to look into it we would gladly take a pull request!

I think the problem is happening in [StaticGeometryColorBatch](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/DataSources/StaticGeometryColorBatch.js)
`StaticGeometryColorBatch` creates one primitive that has geometry used to draw each polygon entity.  When a new polygon entity is added, `StaticGeometryColorBatch` destroys the old primitive and creates a new one with all previous polygons and the new polygon.  
I'm not sure if this is the problem, but to me it looks like for a split second both the old and new primitive are being drawn, and then the old one is removed.

**Comment by roderickgreen** [comment]
@hpinkos thanks for the pointer. I'm not sure how long it will take for me to get up to speed on the source to try to tackle this but I'll at least give it a shot. 

For others, a possible workaround is to group entities in [CustomDataSources](https://cesiumjs.org/Cesium/Build/Documentation/CustomDataSource.html) to control which entities end up in batches for rendering. See [this](https://cesiumjs.org/Cesium/Build/Apps/Sandcastle/#c=pVZ/b9o4GP4qFv80aNSEsmoSY9VRmmnV6K0a3abTcTqZ5CXxzbEj26FjJ777/CMJSUu7Scs/xM/7+7EfhxXfEom2FO5BojeIwz2ag6Jljj87LDiJ3XIuuCaUgzwZoP9XHCFNc2AGmKANYQoGFiOc5kRTwdugAhZb7JonNCZayLaR8o24FN/a0KZkTMUSgF+WWndzpSBikUAnxZooWJAdyFsaf+2aVAwcbkzApB5qWSN4OY/+jM6ufBc5SUHubqXY0kP2Fd/3X6+45wa7XDhlYg3YlpwLJixjVWK3xF/eXd9FJmblWJWQRIzRQoHxq9IA11RTUJgkSeCILISinrM6FZHavBE+xhsp8itIDRkqOB2FYxwO0MsQh33XNie5mezkIyQIqjqCI1XKDYnhxLlU+MTvGUIZ0DTTExQO/FpBTm8oF3L2jaoJOjsPzYM7VvJfbX0Zdqxmq0FSwiZdDj5GV/ie6mzGiowEIR73rfu+YtMzs2Yl3Aq2S02/z1BTDXhpvA1N3v3hhBX+5IQZBUlknO1+wu9MSrIL/vZBCJ2ORueW7fGrZt4WeNYBw1cOHHfBMweOjoEmicf+6T/P5eXiU/QsmcGm5E5dRklKE8beEZ4wI9u+58OSnXmoq+6lU9iyMDxGW8N8Hdc57TExu6V8JVTnwQr0NS9KPXOFDx0EsNX9ehts4bgm+rDHsdlRSXBhpOqlIWhi43CtggE6ojeoXW0rNjvdoKDJ3tQ0F1CSVMdqplsOVdTecTd4ioK7XQF4Eb29+3e+uJ6/r8du5nsieXvihGiyFKWMocv2vFTanLTGGtQt1TSJVJIio3HrQmmh7pw2x/bxYDYLI/oQfEN0hrWor452BWwcqS4T6AQ7Hf5KsODpo2golIkeVXyhFgtHNG2fh5o9ptvf0K59TKOnpq+B5cW9PGV8ccT44peMNsXB1kj5GTn/FS0WH74cEXRzOs1Pw6zXwYFMT+Jh7R33lvV9P/DHtflcfRcivxPBg6vVOvUGvanSOwYXvs4fNC+E1KiULMB4qCEvzHCghuvSfE81jpULmw7roGlCt4gmb1a9B38NVj0UM6KUsdiP+JJ+h1XvYjo0/p0wJkhCefphC5KRnXXJRhcLD2KMp0OzfBylhWBrIlsZfwA) example.

**Comment by roderickgreen** [comment]
FYI I have a fix for this. I need to wait for people to get back from vacation to get the CLA signed, then I'll work to put together a pull request.

**Comment by hpinkos** [comment]
Thanks @roderickgreen!  Looking forward to your contribution =)

**Comment by nswim1** [comment]
Another temporary hack workaround for those waiting for the merge and are seeing this with GroundOverlays in KML:

```
tick = async () => {
  this.kmlDs = await this.telKmlDataSource.load(kmlDocument);
  let rectangle = undefined;
  //find the flickering entity
  for (var entity of this.telKmlDs.entities.values) {
    rectangle = entity;
  }
  rectangle.show = false;
  await this.wait(100);
  rectangle.show = true;
}

wait = ms => new Promise((r)=>setTimeout(r, ms))
```

**Comment by cesium-concierge** [comment]
Congratulations on closing the issue! I found these Cesium forum links in the comments above:

https://groups.google.com/forum/?hl&#x3D;en#!topic/cesium-dev/xKVI49Ux430

If this issue affects any of these threads, please post a comment like the following:

> The issue at https://github.com/AnalyticalGraphicsInc/cesium/issues/3945 has just been closed and may resolve your issue. Look for the change in the next stable release of Cesium or get it now in the master branch on GitHub https://github.com/AnalyticalGraphicsInc/cesium.
---

__I am a bot who helps you make Cesium awesome! [Contributions to my configuration are welcome.](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/.concierge)__

:earth_africa: :earth_americas: :earth_asia:

**Comment by thw0rted** [comment]
FWIW I just ran into this a few weeks before the fix was released -- my workaround was to make a fixed collection of entities whose dynamic properties were CallbackProperties, with the actual by-the-minute values tracked in a separate object.  For example, I had several polygons on the ground represented by a PolylineVolume each.  I got flickering when I dynamically added and removed them (several times per second) but buttery smooth performance when I kept a persistent collection of entities each of which dynamically referenced a corresponding array of Positions representing their hierarchy.

I realize the actual flicker is fixed, but if you find yourself running into performance issues dynamically adding and removing, consider testing if the approach I've described helps.
